app/coffee/WebsocketController.coffee

```diff
@@ -51,6 +54,7 @@ module.exports = WebsocketController =
 	# is determined by FLUSH_IF_EMPTY_DELAY.
 	FLUSH_IF_EMPTY_DELAY: 500 #ms
 	leaveProject: (io, client, callback = (error) ->) ->
+		return callback(new Error('not joined')) unless client.joinedProject
```

Comment:
let's move the metric into the getClientAttributes and add a check on `project_id` instead of adding a new `joinedProject` attribute.  

I think we can skip returning an error and silently return from leaveProject() if the client never joined, that makes the rest of the code simpler.  

If you put the metric after this check then we can see the difference between `editor.leave-project` and `socket.io-disconnect` to tell how many clients are disconnecting without joining.  Alternatively we can have a separate metric for it if you think it's worth tracking.
```
return callback() if !project_id?  # client did not join project
metrics.inc "editor.leave-project"
```

Reply:
This now breaks other unit tests. I also found flawed code that handed the missing user_id and project_id at a later stage. See commit comment:

 - user_id bailout

   There is a check on a completed joinProject call now. It will always
    set a user_id, see Router.coffee which has a fallback `{_id:"..."}`.

 - late project_id bailout

   WebsocketLoadBalancer.emitToRoom will not work without a project_id.
   We have to bail out before the call.